File: analysis/R/ryoe.R
# M2/M3 (Eager & Erickson, Ch.3-4) — Rushing Yards Over Expected in R.
# The book's "over expected" spine: model EXPECTED rush yards from game situation; the residual
# (actual - expected) is the rusher/offense skill signal (RYOE).
#   M2 simple linear regression:   rush_yards ~ yards_to_goal
#   M3 multiple regression:        rush_yards ~ down + distance + yards_to_goal +
#                                               offense_score_diff_start + defense_def_rating
# M3 is the production model: its residual is written back per-play as RYOE.
#
# Reads:  data/gold/plays_model.csv
# Writes: data/results/coef__ryoe__r.csv, metrics__ryoe__r.csv, pred__ryoe__r.csv

source("analysis/R/util_io.R")
suppressPackageStartupMessages({ library(dplyr); library(broom) })
set.seed(42)

FEATURES_M3 <- c("down", "distance", "yards_to_goal",
                 "offense_score_diff_start", "defense_def_rating")

plays <- cfb_read_gold("plays_model")

rush <- plays %>%
  filter(is_rush == 1, !is.na(rush_yards),
         !is.na(down), !is.na(distance), !is.na(yards_to_goal),
         !is.na(offense_score_diff_start))

# Mean-impute the opponent-defense control so every FBS rush keeps a residual (FCS defenses
# have no SP+ rating). Imputing to the mean leaves that row's contribution neutral.
def_mean <- mean(rush$defense_def_rating, na.rm = TRUE)
rush <- rush %>% mutate(defense_def_rating = coalesce(defense_def_rating, def_mean))

rmse <- function(actual, pred) sqrt(mean((actual - pred)^2))

fit_and_record <- function(formula, model_name) {
  m <- lm(formula, data = rush)
  pred <- predict(m, rush)
  list(
    model = m,
    pred = pred,
    coef = tidy(m) %>%
      transmute(model = model_name, term, estimate, std_error = std.error,
                statistic, p_value = p.value, language = "r"),
    metrics = data.frame(
      model = model_name,
      metric = c("rmse", "r_squared", "n_obs"),
      value = c(rmse(rush$rush_yards, pred), summary(m)$r.squared, nrow(rush)),
      language = "r")
  )
}

m2 <- fit_and_record(rush_yards ~ yards_to_goal, "ryoe_simple")
m3 <- fit_and_record(
  as.formula(paste("rush_yards ~", paste(FEATURES_M3, collapse = " + "))),
  "ryoe_multiple")

# baseline = predict the global mean rush_yards for everyone (what RYOE must beat)
base_rmse <- rmse(rush$rush_yards, mean(rush$rush_yards))

coef <- bind_rows(m2$coef, m3$coef)
metrics <- bind_rows(
  m2$metrics, m3$metrics,
  data.frame(model = "baseline_mean", metric = "rmse", value = base_rmse, language = "r"))

# production RYOE = M3 residual, written back per play_key
pred <- data.frame(
  play_key = rush$play_key,
  expected_rush_yards = m3$pred,
  ryoe = rush$rush_yards - m3$pred,
  language = "r")

cfb_write_result("coef__ryoe__r", coef)
cfb_write_result("metrics__ryoe__r", metrics)
cfb_write_result("pred__ryoe__r", pred)
cat(sprintf("[ryoe/R] M3 RMSE=%.3f (baseline %.3f), R2=%.4f, %d plays\n",
            m3$metrics$value[1], base_rmse, m3$metrics$value[2], nrow(rush)))

File: analysis/python/ryoe.py
"""M2/M3 (Eager & Erickson, Ch.3-4) — Rushing Yards Over Expected, Python parity of ryoe.R.

statsmodels OLS mirrors R's lm() exactly (same design matrix, same closed-form fit), so the
coefficients must agree to numerical tolerance — that agreement is the committed correctness
check. Same feed, features, mean-imputation, and filtering as the R version.

Reads data/gold/plays_model.csv; writes coef__ryoe__py / metrics__ryoe__py / pred__ryoe__py.
"""
from __future__ import annotations

import numpy as np
import pandas as pd
import statsmodels.formula.api as smf

from cfb_analytics.config import REPO_ROOT

GOLD = REPO_ROOT / "data" / "gold"
RESULTS = REPO_ROOT / "data" / "results"
FEATURES_M3 = ["down", "distance", "yards_to_goal",
               "offense_score_diff_start", "defense_def_rating"]

# statsmodels names the intercept "Intercept"; R names it "(Intercept)". Normalise to R's so the
# parity join lines terms up.
_TERM = {"Intercept": "(Intercept)"}


def _rmse(actual: np.ndarray, pred: np.ndarray) -> float:
    return float(np.sqrt(np.mean((actual - pred) ** 2)))


def _coef_frame(res, model_name: str) -> pd.DataFrame:
    return pd.DataFrame({
        "model": model_name,
        "term": [_TERM.get(t, t) for t in res.params.index],
        "estimate": res.params.values,
        "std_error": res.bse.values,
        "statistic": res.tvalues.values,
        "p_value": res.pvalues.values,
        "language": "py",
    })


def main() -> None:
    plays = pd.read_csv(GOLD / "plays_model.csv")
    rush = plays[(plays.is_rush == 1) & plays.rush_yards.notna()
                 & plays.down.notna() & plays.distance.notna()
                 & plays.yards_to_goal.notna()
                 & plays.offense_score_diff_start.notna()].copy()

    def_mean = rush["defense_def_rating"].mean()
    rush["defense_def_rating"] = rush["defense_def_rating"].fillna(def_mean)

    m2 = smf.ols("rush_yards ~ yards_to_goal", data=rush).fit()
    m3 = smf.ols("rush_yards ~ " + " + ".join(FEATURES_M3), data=rush).fit()

    pred3 = m3.predict(rush).to_numpy()
    actual = rush["rush_yards"].to_numpy()
    base_rmse = _rmse(actual, np.full_like(actual, actual.mean(), dtype=float))

    coef = pd.concat([_coef_frame(m2, "ryoe_simple"),
                      _coef_frame(m3, "ryoe_multiple")], ignore_index=True)
    metrics = pd.DataFrame([
        dict(model="ryoe_simple", metric="rmse", value=_rmse(actual, m2.predict(rush))),
        dict(model="ryoe_simple", metric="r_squared", value=m2.rsquared),
        dict(model="ryoe_multiple", metric="rmse", value=_rmse(actual, pred3)),
        dict(model="ryoe_multiple", metric="r_squared", value=m3.rsquared),
        dict(model="ryoe_multiple", metric="n_obs", value=float(len(rush))),
        dict(model="baseline_mean", metric="rmse", value=base_rmse),
    ])
    metrics["language"] = "py"

    pred = pd.DataFrame({
        "play_key": rush["play_key"].values,
        "expected_rush_yards": pred3,
        "ryoe": actual - pred3,
        "language": "py",
    })

    RESULTS.mkdir(parents=True, exist_ok=True)
    coef.to_csv(RESULTS / "coef__ryoe__py.csv", index=False)
    metrics.to_csv(RESULTS / "metrics__ryoe__py.csv", index=False)
    pred.to_csv(RESULTS / "pred__ryoe__py.csv", index=False)
    print(f"[ryoe/py] M3 RMSE={_rmse(actual, pred3):.3f} (baseline {base_rmse:.3f}), "
          f"R2={m3.rsquared:.4f}, {len(rush)} plays")


if __name__ == "__main__":
    main()
